### DISCUSSION TITLE
Distance Record for each ship - Postgre SQL

### DISCUSSION THREAD

**Discussion by rkarikari** [jvde-github/AIS-catcher#145] [General]
Hello ,
The idea was is to use a copy of  database table ("ais_vessel") that for each ship/mmsi captures the last transmission.

If the ship is leaving, the last packet received if effectively the distance record for that ship.
If it's approaching, then further updates will lower this distance... i.e update ("record") only if last packet distance is lower than new packet .

I created a new table ("record") for this . ....... and insert/update rows only if distance of packet is greater.
A trigger function pushes any updates/insertions to the new table.

[TriggerT.txt](https://github.com/jvde-github/AIS-catcher/files/12378441/TriggerT.txt)

I'm an absolute NEWbie in sql so any help/ideas/optimisation would be  greatly appreciated.

Thanks in advance.

**Comment by jvde-github** [comment]
Interesting concept and nice idea. Unfortunately, the distances are not available as database output as they are calculated in a different part of the code. So this is a nice solution with triggers and distance calc in the DB. I have not tested it, does it work well?

You could make a table as well with only mmsi and record distance and not replicate the other fields?

**Comment by rkarikari** [comment]
<img width="791" alt="recordx" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/e032b67d-6227-455a-8d60-24b23da7ac14">
<img width="790" alt="ais_vessel" src="https://github.com/jvde-github/AIS-catcher/assets/115777368/1c779fd2-4eb9-4c96-9e6a-4134fcf1da7d">

The highlighted SHIP/Vessel      ....          MSC ANNAMARIA

came in at                                                                  -   08:16 am           46.26Nm (record) 
but was last heard from at                                       -   10:16 am            36.66Nm   away. 

Now I'll have the info on which ships are pushing the boundaries of the coverage area.... I've always wondered about that.

**Comment by jvde-github** [comment]
Really nice! Love it and quite elegant. Not sure how efficient triggers are but better than saving all path data

**Comment by rkarikari** [comment]
> You could make a table as well with only mmsi and record distance and not replicate the other fields?

Yes, that's true !!

**Reply by jvde-github** [reply]
But now you see when the record was set.... I could capture it in the AIS-catcher web screen as well. Will not be in the database so your solution works for that, but if it is of interest to people we can add it there as well. Had not thought about it before.

**Reply by jvde-github** [reply]
Btw, can you also add a column to the original ais_vessel_table and store the record there?

**Comment by rkarikari** [comment]
> Btw, can you also add a column to the original ais_vessel_table and store the record there?

I was toying with multiple ideas.
I'm currently not storing the data... it's a  calculated field from the gps coordinates.

You should definitely add this feature in any way possible.  I'm pretty sure it will be greatly appreciated.

**Reply by jvde-github** [reply]
Nice idea. I thought about using stored procedures so users can more easily tweak how the database is set up by modifying the database and not the AIS-catcher code. Was a bit hesitant as it makes things more complex as well. Your trigger idea I did not think of.

**Comment by rkarikari** [comment]
I don't have enough traffic in my area to test  ... performance wise.
So hopefully the experts  :)   will do it  better.

AIS-catcher is great!   thanks so much for all the work done.

My interest  in Amateur radio just got a boost    ...and the ships serve as  range test buddies. 
:)

**Comment by mdkendall** [comment]
You don't actually need to divide by 1852 in the WHERE clause (you are just comparing numbers and it doesn't matter if they are in nautical miles or not) which might make a small performance improvement.

**Reply by rkarikari** [reply]
Of course you are right ! 

Just some  "cobwebs" missed  I guess  Thank you

**Comment by rkarikari** [comment]
Hmm...
I wonder ....  It reduces the frequency of updates in it's current form.
I would not want updates on centimeter changes.  Wish I could test somehow.
